ZOOKEEPER-1932: Remove deprecated LeaderElection class.

hanm opened the pull request:
The motivation of removing LeaderElection class:
* It has been long deprecated and no one uses it.
* Tests around it is flaky.

lvfangmin commented:
Since we're deprecating the LeaderElection, should we also remove LeaderElection.java itself? Any reason we keep it there even no code is referencing it?

hanm commented:
@lvfangmin Good catch, thanks for looking at the PR. I can't remember why LeaderElection.java was not removed in the patch, let me double check before updating the PR.

hanm commented:
PR updated:

* Removed LeaderElection class that not referenced anywhere.
* Fixed a couple of places in tests that still reference the LeaderElection algorithm. 

Would be good to get this in as this LETest was one of most frequent flaky tests in recent builds. @rgs1 @rakeshadr @fpj if you have cycles please review and merge if it's lgty.

hanm commented:
Need another committer view to merge this in; @arshadmohammad, can you please review this? Flavio had +1 on an older version of the patch as commented in JIRA, and this is only going to trunk.

arshadmohammad commented:
I will review it soon

hanm commented:
Thanks @arshadmohammad

arshadmohammad commented on src/docs/src/documentation/content/xdocs/zookeeperAdmin.xml:
```diff
@@ -948,15 +948,14 @@ server.3=zoo3:2888:3888</programlisting>
+              <para> The implementations of leader election 1, and 2 are now
               <emphasis role="bold"> deprecated </emphasis>. We have the intention
               of removing them in the next release, at which point only the 
               FastLeaderElection will be available. 
```
There is at least one more reference of election alog type 0. Should be deleted.
Reference Text : If electionAlg is 0, then the second port is not 
              necessary

hanm replied:
removed.

arshadmohammad commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -930,39 +918,36 @@ protected Election createElectionAlgorithm(int electionAlgorithm){
-        }
         return electionAlg;
     }
 
```
org.apache.zookeeper.server.quorum.QuorumPeerConfig.parseDynamicConfig(Properties, int, boolean, boolean)
has a logic related to election algo type 0. This should be removed

hanm replied:
good catch. fixed.

arshadmohammad commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -809,28 +809,16 @@ synchronized public void stopLeaderElection() {
         responder.interrupt();
```
The import java.net.SocketException is never used

hanm replied:
removed.

arshadmohammad commented:
Jira ZOOKEEPER-1932 title is completely different from what is being done in its MR. It think we should change the title as well as the description

arshadmohammad commented on src/java/main/org/apache/zookeeper/server/quorum/QuorumPeer.java:
```diff
@@ -930,39 +918,36 @@ protected Election createElectionAlgorithm(int electionAlgorithm){
-        default:
-            assert false;
+            case 0:
+                assert false : "Leader election algorithm type 0 is not supported anymore.";
```
I think we should log error here and make sure the server jvm exit in this case

arshadmohammad replied:
Another option is  validate electionAlg in org.apache.zookeeper.server.quorum.QuorumPeerConfig and exit from there for wrong input and remove the case from here

hanm replied:
I like the idea of catching invalid electionAlg value when parsing config file. Code updated by throwing an exception when input value is bad.

hanm commented:
@arshadmohammad thanks for review, code is updated, please take a look at. I also updated JIRA description so it's consistent with the pull request title.

arshadmohammad commented:
LGTM  +1
